File: src/test/java/com/customagent/service/AnalysisServiceTest.java
package com.customagent.service;

import com.customagent.exception.AgentNotFoundException;
import com.customagent.exception.AnalysisException;
import com.customagent.model.Agent;
import com.customagent.model.AgentStatus;
import com.customagent.model.AnalysisReport;
import com.customagent.model.ReportStatus;
import com.customagent.model.dto.AnalysisRequest;
import com.customagent.repository.AgentRepository;
import com.customagent.repository.AnalysisReportRepository;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.Assertions.*;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.*;

/**
 * Unit tests for {@link AnalysisService}.
 *
 * <p>Both {@link AgentRepository} and {@link AnalysisReportRepository} are mocked.
 * State-transition logic (PENDING → COMPLETED / FAILED) is exercised exhaustively.
 */
@ExtendWith(MockitoExtension.class)
class AnalysisServiceTest {

    @Mock  private AnalysisReportRepository reportRepository;
    @Mock  private AgentRepository          agentRepository;

    @InjectMocks
    private AnalysisService analysisService;

    // ── helpers ──────────────────────────────────────────────────────────────

    private Agent activeAgent(Long id) {
        return Agent.builder().id(id).name("Agent-" + id)
                .status(AgentStatus.ACTIVE).build();
    }

    private Agent inactiveAgent(Long id) {
        return Agent.builder().id(id).name("Agent-" + id)
                .status(AgentStatus.INACTIVE).build();
    }

    private AnalysisReport reportWithStatus(Long id, ReportStatus status) {
        return AnalysisReport.builder()
                .id(id)
                .repositoryUrl("https://github.com/test/repo")
                .agentId(1L)
                .status(status)
                .build();
    }

    // =========================================================================
    @Nested
    @DisplayName("triggerAnalysis()")
    class TriggerAnalysis {

        @Test
        @DisplayName("should create PENDING report when agent is active")
        void whenAgentIsActive_shouldCreatePendingReport() {
            Agent agent = activeAgent(1L);
            AnalysisRequest request = new AnalysisRequest("https://github.com/org/repo", 1L);
            AnalysisReport saved = reportWithStatus(10L, ReportStatus.PENDING);

            when(agentRepository.findById(1L)).thenReturn(Optional.of(agent));
            when(reportRepository.save(any(AnalysisReport.class))).thenReturn(saved);

            AnalysisReport result = analysisService.triggerAnalysis(request);

            assertThat(result.getStatus()).isEqualTo(ReportStatus.PENDING);
            assertThat(result.getId()).isEqualTo(10L);

            // Verify the persisted object has the correct fields
            ArgumentCaptor<AnalysisReport> captor =
                    ArgumentCaptor.forClass(AnalysisReport.class);
            verify(reportRepository).save(captor.capture());
            AnalysisReport toSave = captor.getValue();
            assertThat(toSave.getRepositoryUrl()).isEqualTo("https://github.com/org/repo");
            assertThat(toSave.getAgentId()).isEqualTo(1L);
            assertThat(toSave.getStatus()).isEqualTo(ReportStatus.PENDING);
        }

        @Test
        @DisplayName("should throw AgentNotFoundException when agent does not exist")
        void whenAgentNotFound_shouldThrowAgentNotFoundException() {
            when(agentRepository.findById(99L)).thenReturn(Optional.empty());
            AnalysisRequest request = new AnalysisRequest("https://github.com/test", 99L);

            assertThatThrownBy(() -> analysisService.triggerAnalysis(request))
                    .isInstanceOf(AgentNotFoundException.class)
                    .hasMessageContaining("99");

            verify(reportRepository, never()).save(any());
        }

        @Test
        @DisplayName("should throw AnalysisException when agent is inactive")
        void whenAgentIsInactive_shouldThrowAnalysisException() {
            Agent agent = inactiveAgent(2L);
            when(agentRepository.findById(2L)).thenReturn(Optional.of(agent));
            AnalysisRequest request = new AnalysisRequest("https://github.com/test", 2L);

            assertThatThrownBy(() -> analysisService.triggerAnalysis(request))
                    .isInstanceOf(AnalysisException.class)
                    .hasMessageContaining("inactive");

            verify(reportRepository, never()).save(any());
        }

        @Test
        @DisplayName("saved report should carry the resolved agent's id")
        void whenAgentIsActive_savedReportShouldCarryAgentId() {
            Agent agent = activeAgent(7L);
            AnalysisRequest request = new AnalysisRequest("https://github.com/org/repo", 7L);
            AnalysisReport saved = reportWithStatus(20L, ReportStatus.PENDING);

            when(agentRepository.findById(7L)).thenReturn(Optional.of(agent));
            when(reportRepository.save(any(AnalysisReport.class))).thenReturn(saved);

            analysisService.triggerAnalysis(request);

            ArgumentCaptor<AnalysisReport> captor =
                    ArgumentCaptor.forClass(AnalysisReport.class);
            verify(reportRepository).save(captor.capture());
            assertThat(captor.getValue().getAgentId()).isEqualTo(7L);
            assertThat(captor.getValue().getRepositoryUrl())
                    .isEqualTo("https://github.com/org/repo");
        }
    }

    // =========================================================================
    @Nested
    @DisplayName("getReportById()")
    class GetReportById {

        @Test
        @DisplayName("should return report when it exists")
        void whenReportExists_shouldReturnReport() {
            AnalysisReport report = reportWithStatus(5L, ReportStatus.COMPLETED);
            when(reportRepository.findById(5L)).thenReturn(Optional.of(report));

            AnalysisReport result = analysisService.getReportById(5L);

            assertThat(result.getId()).isEqualTo(5L);
            assertThat(result.getStatus()).isEqualTo(ReportStatus.COMPLETED);
        }

        @Test
        @DisplayName("should throw AnalysisException when report does not exist")
        void whenReportNotFound_shouldThrowAnalysisException() {
            when(reportRepository.findById(404L)).thenReturn(Optional.empty());

            assertThatThrownBy(() -> analysisService.getReportById(404L))
                    .isInstanceOf(AnalysisException.class)
                    .hasMessageContaining("404");
        }
    }

    // =========================================================================
    @Nested
    @DisplayName("getReportsByAgentId()")
    class GetReportsByAgentId {

        @Test
        @DisplayName("should return all reports for the given agent")
        void shouldReturnReportsForAgent() {
            List<AnalysisReport> reports = List.of(
                    reportWithStatus(1L, ReportStatus.PENDING),
                    reportWithStatus(2L, ReportStatus.COMPLETED));
            when(reportRepository.findByAgentId(7L)).thenReturn(reports);

            assertThat(analysisService.getReportsByAgentId(7L)).hasSize(2);
        }

        @Test
        @DisplayName("should return empty list when agent has no reports")
        void whenNoReports_shouldReturnEmptyList() {
            when(reportRepository.findByAgentId(7L)).thenReturn(List.of());

            assertThat(analysisService.getReportsByAgentId(7L)).isEmpty();
        }
    }

    // =========================================================================
    @Nested
    @DisplayName("getReportsByStatus()")
    class GetReportsByStatus {

        @Test
        @DisplayName("should delegate to repository and return matching reports")
        void shouldReturnMatchingReports() {
            List<AnalysisReport> reports =
                    List.of(reportWithStatus(1L, ReportStatus.FAILED));
            when(reportRepository.findByStatus(ReportStatus.FAILED)).thenReturn(reports);

            assertThat(analysisService.getReportsByStatus(ReportStatus.FAILED))
                    .hasSize(1)
                    .allMatch(r -> r.getStatus() == ReportStatus.FAILED);
        }

        @Test
        @DisplayName("should return empty list when no reports match the given status")
        void whenNoMatchingReports_shouldReturnEmptyList() {
            when(reportRepository.findByStatus(ReportStatus.PENDING)).thenReturn(List.of());

            assertThat(analysisService.getReportsByStatus(ReportStatus.PENDING)).isEmpty();
        }

        @Test
        @DisplayName("should return COMPLETED reports when queried by COMPLETED status")
        void forCompletedStatus_shouldReturnMatchingReports() {
            List<AnalysisReport> reports =
                    List.of(reportWithStatus(3L, ReportStatus.COMPLETED),
                            reportWithStatus(4L, ReportStatus.COMPLETED));
            when(reportRepository.findByStatus(ReportStatus.COMPLETED)).thenReturn(reports);

            assertThat(analysisService.getReportsByStatus(ReportStatus.COMPLETED))
                    .hasSize(2)
                    .allMatch(r -> r.getStatus() == ReportStatus.COMPLETED);
        }

        @Test
        @DisplayName("should return IN_PROGRESS reports when queried by IN_PROGRESS status")
        void forInProgressStatus_shouldReturnMatchingReports() {
            List<AnalysisReport> reports =
                    List.of(reportWithStatus(5L, ReportStatus.IN_PROGRESS));
            when(reportRepository.findByStatus(ReportStatus.IN_PROGRESS)).thenReturn(reports);

            assertThat(analysisService.getReportsByStatus(ReportStatus.IN_PROGRESS))
                    .hasSize(1)
                    .allMatch(r -> r.getStatus() == ReportStatus.IN_PROGRESS);
        }
    }

    // =========================================================================
    @Nested
    @DisplayName("completeAnalysis()")
    class CompleteAnalysis {

        @Test
        @DisplayName("should transition PENDING → COMPLETED and store result")
        void whenPending_shouldCompleteWithResult() {
            AnalysisReport report = reportWithStatus(1L, ReportStatus.PENDING);
            when(reportRepository.findById(1L)).thenReturn(Optional.of(report));
            when(reportRepository.save(any())).thenAnswer(inv -> inv.getArgument(0));

            AnalysisReport result = analysisService.completeAnalysis(1L, "{\"tests\":5}");

            assertThat(result.getStatus()).isEqualTo(ReportStatus.COMPLETED);
            assertThat(result.getResult()).isEqualTo("{\"tests\":5}");
            assertThat(result.getCompletedAt()).isNotNull();
            assertThat(result.getCompletedAt()).isBeforeOrEqualTo(LocalDateTime.now());
        }

        @Test
        @DisplayName("should transition IN_PROGRESS → COMPLETED")
        void whenInProgress_shouldCompleteWithResult() {
            AnalysisReport report = reportWithStatus(2L, ReportStatus.IN_PROGRESS);
            when(reportRepository.findById(2L)).thenReturn(Optional.of(report));
            when(reportRepository.save(any())).thenAnswer(inv -> inv.getArgument(0));

            AnalysisReport result = analysisService.completeAnalysis(2L, "ok");

            assertThat(result.getStatus()).isEqualTo(ReportStatus.COMPLETED);
        }

        @Test
        @DisplayName("should throw AnalysisException when report is already COMPLETED")
        void whenAlreadyCompleted_shouldThrowAnalysisException() {
            AnalysisReport report = reportWithStatus(3L, ReportStatus.COMPLETED);
            when(reportRepository.findById(3L)).thenReturn(Optional.of(report));

            assertThatThrownBy(() -> analysisService.completeAnalysis(3L, "re-done"))
                    .isInstanceOf(AnalysisException.class)
                    .hasMessageContaining("already completed");

            verify(reportRepository, never()).save(any());
        }

        @Test
        @DisplayName("should throw AnalysisException when report is FAILED")
        void whenFailed_shouldThrowAnalysisException() {
            AnalysisReport report = reportWithStatus(4L, ReportStatus.FAILED);
            when(reportRepository.findById(4L)).thenReturn(Optional.of(report));

            assertThatThrownBy(() -> analysisService.completeAnalysis(4L, "result"))
                    .isInstanceOf(AnalysisException.class)
                    .hasMessageContaining("failed report");
        }

        @Test
        @DisplayName("should throw AnalysisException when report is not found")
        void whenReportNotFound_shouldThrowAnalysisException() {
            when(reportRepository.findById(99L)).thenReturn(Optional.empty());

            assertThatThrownBy(() -> analysisService.completeAnalysis(99L, "result"))
                    .isInstanceOf(AnalysisException.class)
                    .hasMessageContaining("99");

            verify(reportRepository, never()).save(any());
        }

        @Test
        @DisplayName("should call save exactly once when completing a PENDING report")
        void whenPending_shouldCallSaveExactlyOnce() {
            AnalysisReport report = reportWithStatus(10L, ReportStatus.PENDING);
            when(reportRepository.findById(10L)).thenReturn(Optional.of(report));
            when(reportRepository.save(any())).thenAnswer(inv -> inv.getArgument(0));

            analysisService.completeAnalysis(10L, "done");

            verify(reportRepository, times(1)).save(report);
        }

        @Test
        @DisplayName("should accept a null result payload when completing analysis")
        void whenCompletingWithNullResult_shouldSetNullResultAndComplete() {
            AnalysisReport report = reportWithStatus(11L, ReportStatus.PENDING);
            when(reportRepository.findById(11L)).thenReturn(Optional.of(report));
            when(reportRepository.save(any())).thenAnswer(inv -> inv.getArgument(0));

            AnalysisReport result = analysisService.completeAnalysis(11L, null);

            assertThat(result.getStatus()).isEqualTo(ReportStatus.COMPLETED);
            assertThat(result.getResult()).isNull();
            assertThat(result.getCompletedAt()).isNotNull();
        }
    }

    // =========================================================================
    @Nested
    @DisplayName("failAnalysis()")
    class FailAnalysis {

        @Test
        @DisplayName("should transition PENDING → FAILED and store error message")
        void whenPending_shouldFailWithMessage() {
            AnalysisReport report = reportWithStatus(1L, ReportStatus.PENDING);
            when(reportRepository.findById(1L)).thenReturn(Optional.of(report));
            when(reportRepository.save(any())).thenAnswer(inv -> inv.getArgument(0));

            AnalysisReport result = analysisService.failAnalysis(1L, "timeout");

            assertThat(result.getStatus()).isEqualTo(ReportStatus.FAILED);
            assertThat(result.getResult()).isEqualTo("timeout");
            assertThat(result.getCompletedAt()).isNotNull();
        }

        @Test
        @DisplayName("should transition IN_PROGRESS → FAILED")
        void whenInProgress_shouldFailWithMessage() {
            AnalysisReport report = reportWithStatus(2L, ReportStatus.IN_PROGRESS);
            when(reportRepository.findById(2L)).thenReturn(Optional.of(report));
            when(reportRepository.save(any())).thenAnswer(inv -> inv.getArgument(0));

            AnalysisReport result = analysisService.failAnalysis(2L, "error");

            assertThat(result.getStatus()).isEqualTo(ReportStatus.FAILED);
        }

        @Test
        @DisplayName("should throw AnalysisException when report is already COMPLETED")
        void whenCompleted_shouldThrowAnalysisException() {
            AnalysisReport report = reportWithStatus(3L, ReportStatus.COMPLETED);
            when(reportRepository.findById(3L)).thenReturn(Optional.of(report));

            assertThatThrownBy(() -> analysisService.failAnalysis(3L, "err"))
                    .isInstanceOf(AnalysisException.class)
                    .hasMessageContaining("terminal status");
        }

        @Test
        @DisplayName("should throw AnalysisException when report is already FAILED")
        void whenAlreadyFailed_shouldThrowAnalysisException() {
            AnalysisReport report = reportWithStatus(4L, ReportStatus.FAILED);
            when(reportRepository.findById(4L)).thenReturn(Optional.of(report));

            assertThatThrownBy(() -> analysisService.failAnalysis(4L, "err"))
                    .isInstanceOf(AnalysisException.class)
                    .hasMessageContaining("terminal status");
        }

        @Test
        @DisplayName("should throw AnalysisException when report is not found")
        void whenReportNotFound_shouldThrowAnalysisException() {
            when(reportRepository.findById(99L)).thenReturn(Optional.empty());

            assertThatThrownBy(() -> analysisService.failAnalysis(99L, "error"))
                    .isInstanceOf(AnalysisException.class)
                    .hasMessageContaining("99");

            verify(reportRepository, never()).save(any());
        }

        @Test
        @DisplayName("should call save exactly once when failing a PENDING report")
        void whenPending_shouldCallSaveExactlyOnce() {
            AnalysisReport report = reportWithStatus(10L, ReportStatus.PENDING);
            when(reportRepository.findById(10L)).thenReturn(Optional.of(report));
            when(reportRepository.save(any())).thenAnswer(inv -> inv.getArgument(0));

            analysisService.failAnalysis(10L, "connection timeout");

            verify(reportRepository, times(1)).save(report);
        }

        @Test
        @DisplayName("completedAt should be set when transitioning IN_PROGRESS → FAILED")
        void whenInProgress_completedAtShouldBeSet() {
            AnalysisReport report = reportWithStatus(11L, ReportStatus.IN_PROGRESS);
            when(reportRepository.findById(11L)).thenReturn(Optional.of(report));
            when(reportRepository.save(any())).thenAnswer(inv -> inv.getArgument(0));

            AnalysisReport result = analysisService.failAnalysis(11L, "out-of-memory");

            assertThat(result.getCompletedAt())
                    .isNotNull()
                    .isBeforeOrEqualTo(LocalDateTime.now());
        }
    }
}

File: src/test/analysisService.test.js
'use strict';

/**
 * Unit tests for src/services/analysisService.js
 * Framework: Jest (CommonJS)
 * Coverage targets: statements ≥80%, branches ≥70%
 */

const {
  createAnalysisService,
  ReportStatus,
  resetIdCounter,
} = require('../services/analysisService');

// ── Repository factories ──────────────────────────────────────────────────────

function makeReportRepository(initialReports = []) {
  const store = new Map(initialReports.map((r) => [r.id, { ...r }]));

  return {
    findById:      jest.fn((id)       => store.get(id) || null),
    findByAgentId: jest.fn((agentId)  => [...store.values()].filter((r) => r.agentId === agentId)),
    findByStatus:  jest.fn((status)   => [...store.values()].filter((r) => r.status === status)),
    save:          jest.fn((report)   => { store.set(report.id, report); return report; }),
    _store: store,
  };
}

function makeAgentRepository(agents = []) {
  const store = new Map(agents.map((a) => [a.id, { ...a }]));
  return {
    findById: jest.fn((id) => store.get(id) || null),
    _store: store,
  };
}

// ── Helpers ───────────────────────────────────────────────────────────────────

const ACTIVE_AGENT   = { id: 1, name: 'ActiveAgent',   status: 'ACTIVE'   };
const INACTIVE_AGENT = { id: 2, name: 'InactiveAgent', status: 'INACTIVE' };

function makeService(reports = [], agentList = [ACTIVE_AGENT]) {
  const reportRepo = makeReportRepository(reports);
  const agentRepo  = makeAgentRepository(agentList);
  return { service: createAnalysisService(reportRepo, agentRepo), reportRepo, agentRepo };
}

function makeReport(overrides = {}) {
  return {
    id: 1,
    repositoryUrl: 'https://github.com/org/repo',
    agentId: 1,
    status: ReportStatus.PENDING,
    result: null,
    createdAt: new Date().toISOString(),
    completedAt: null,
    ...overrides,
  };
}

// ─────────────────────────────────────────────────────────────────────────────
describe('AnalysisService — triggerAnalysis', () => {
  beforeEach(() => resetIdCounter(1));

  it('should create and return a PENDING report when agent is ACTIVE', () => {
    // Arrange
    const { service } = makeService([], [ACTIVE_AGENT]);
    const request = { repositoryUrl: 'https://github.com/org/repo', agentId: 1 };

    // Act
    const result = service.triggerAnalysis(request);

    // Assert
    expect(result.status).toBe(ReportStatus.PENDING);
    expect(result.repositoryUrl).toBe('https://github.com/org/repo');
    expect(result.agentId).toBe(1);
    expect(result.id).toBeDefined();
    expect(result.completedAt).toBeNull();
  });

  it('should persist the report via reportRepository.save', () => {
    // Arrange
    const { service, reportRepo } = makeService([], [ACTIVE_AGENT]);

    // Act
    service.triggerAnalysis({ repositoryUrl: 'https://github.com/a/b', agentId: 1 });

    // Assert
    expect(reportRepo.save).toHaveBeenCalledTimes(1);
    expect(reportRepo.save).toHaveBeenCalledWith(
      expect.objectContaining({ status: ReportStatus.PENDING, agentId: 1 })
    );
  });

  it('should trim whitespace from repositoryUrl', () => {
    // Arrange
    const { service } = makeService([], [ACTIVE_AGENT]);

    // Act
    const result = service.triggerAnalysis({ repositoryUrl: '  https://github.com/a/b  ', agentId: 1 });

    // Assert
    expect(result.repositoryUrl).toBe('https://github.com/a/b');
  });

  // ── Validation failures ────────────────────────────────────────────────────

  it('should throw when repositoryUrl is missing', () => {
    const { service } = makeService([], [ACTIVE_AGENT]);
    expect(() => service.triggerAnalysis({ agentId: 1 })).toThrow('Repository URL is required');
  });

  it('should throw when repositoryUrl is empty string', () => {
    const { service } = makeService([], [ACTIVE_AGENT]);
    expect(() => service.triggerAnalysis({ repositoryUrl: '', agentId: 1 })).toThrow(
      'Repository URL is required'
    );
  });

  it('should throw when agentId is null', () => {
    const { service } = makeService([], [ACTIVE_AGENT]);
    expect(() =>
      service.triggerAnalysis({ repositoryUrl: 'https://x.com', agentId: null })
    ).toThrow('Agent ID is required');
  });

  it('should throw when agentId is undefined', () => {
    const { service } = makeService([], [ACTIVE_AGENT]);
    expect(() =>
      service.triggerAnalysis({ repositoryUrl: 'https://x.com' })
    ).toThrow('Agent ID is required');
  });

  it('should throw when request is null', () => {
    const { service } = makeService([], [ACTIVE_AGENT]);
    expect(() => service.triggerAnalysis(null)).toThrow('Repository URL is required');
  });

  // ── Agent state failures ───────────────────────────────────────────────────

  it('should throw when agent is not found', () => {
    const { service } = makeService([], []);
    expect(() =>
      service.triggerAnalysis({ repositoryUrl: 'https://x.com', agentId: 999 })
    ).toThrow('Agent not found with id: 999');
  });

  it('should throw when agent is INACTIVE', () => {
    const { service } = makeService([], [INACTIVE_AGENT]);
    expect(() =>
      service.triggerAnalysis({ repositoryUrl: 'https://x.com', agentId: 2 })
    ).toThrow('Cannot trigger analysis with inactive agent: 2');
  });
});

// ─────────────────────────────────────────────────────────────────────────────
describe('AnalysisService — getReportById', () => {
  it('should return the report when found', () => {
    // Arrange
    const report = makeReport({ id: 7 });
    const { service, reportRepo } = makeService([report]);

    // Act
    const result = service.getReportById(7);

    // Assert
    expect(result).toMatchObject({ id: 7, status: ReportStatus.PENDING });
    expect(reportRepo.findById).toHaveBeenCalledWith(7);
  });

  it('should throw when report is not found', () => {
    const { service } = makeService([]);
    expect(() => service.getReportById(404)).toThrow(
      'Analysis report not found with id: 404'
    );
  });
});

// ─────────────────────────────────────────────────────────────────────────────
describe('AnalysisService — getReportsByAgentId', () => {
  it('should return all reports for an agent', () => {
    // Arrange
    const r1 = makeReport({ id: 1, agentId: 1 });
    const r2 = makeReport({ id: 2, agentId: 1 });
    const r3 = makeReport({ id: 3, agentId: 2 });
    const { service, reportRepo } = makeService([r1, r2, r3]);

    // Act
    const result = service.getReportsByAgentId(1);

    // Assert
    expect(result).toHaveLength(2);
    expect(reportRepo.findByAgentId).toHaveBeenCalledWith(1);
  });

  it('should return empty array when agent has no reports', () => {
    const { service } = makeService([]);
    expect(service.getReportsByAgentId(99)).toEqual([]);
  });
});

// ─────────────────────────────────────────────────────────────────────────────
describe('AnalysisService — getReportsByStatus', () => {
  it('should return reports matching given status', () => {
    // Arrange
    const r1 = makeReport({ id: 1, status: ReportStatus.PENDING   });
    const r2 = makeReport({ id: 2, status: ReportStatus.COMPLETED });
    const r3 = makeReport({ id: 3, status: ReportStatus.PENDING   });
    const { service, reportRepo } = makeService([r1, r2, r3]);

    // Act
    const result = service.getReportsByStatus(ReportStatus.PENDING);

    // Assert
    expect(result).toHaveLength(2);
    expect(reportRepo.findByStatus).toHaveBeenCalledWith(ReportStatus.PENDING);
  });
});

// ─────────────────────────────────────────────────────────────────────────────
describe('AnalysisService — completeAnalysis', () => {
  it('should transition PENDING → COMPLETED and store result', () => {
    // Arrange
    const pending = makeReport({ id: 1, status: ReportStatus.PENDING });
    const { service, reportRepo } = makeService([pending]);

    // Act
    const result = service.completeAnalysis(1, '{"coverage": 85}');

    // Assert
    expect(result.status).toBe(ReportStatus.COMPLETED);
    expect(result.result).toBe('{"coverage": 85}');
    expect(result.completedAt).not.toBeNull();
    expect(reportRepo.save).toHaveBeenCalledTimes(1);
  });

  it('should throw when report is already COMPLETED', () => {
    // Arrange
    const completed = makeReport({ id: 2, status: ReportStatus.COMPLETED });
    const { service } = makeService([completed]);

    // Act & Assert
    expect(() => service.completeAnalysis(2, 'result')).toThrow(
      'Report 2 is already completed'
    );
  });

  it('should throw when report is in FAILED state', () => {
    // Arrange
    const failed = makeReport({ id: 3, status: ReportStatus.FAILED });
    const { service } = makeService([failed]);

    // Act & Assert
    expect(() => service.completeAnalysis(3, 'result')).toThrow(
      'Cannot complete a failed report: 3'
    );
  });

  it('should throw when report does not exist', () => {
    const { service } = makeService([]);
    expect(() => service.completeAnalysis(999, 'result')).toThrow(
      'Analysis report not found with id: 999'
    );
  });
});

// ─────────────────────────────────────────────────────────────────────────────
describe('AnalysisService — failAnalysis', () => {
  it('should transition PENDING → FAILED and store error message', () => {
    // Arrange
    const pending = makeReport({ id: 1, status: ReportStatus.PENDING });
    const { service, reportRepo } = makeService([pending]);

    // Act
    const result = service.failAnalysis(1, 'Connection timeout');

    // Assert
    expect(result.status).toBe(ReportStatus.FAILED);
    expect(result.result).toBe('Connection timeout');
    expect(result.completedAt).not.toBeNull();
    expect(reportRepo.save).toHaveBeenCalledTimes(1);
  });

  it('should throw when report is already COMPLETED', () => {
    // Arrange
    const completed = makeReport({ id: 4, status: ReportStatus.COMPLETED });
    const { service } = makeService([completed]);

    // Act & Assert
    expect(() => service.failAnalysis(4, 'err')).toThrow(
      `Cannot fail report 4 in terminal status: ${ReportStatus.COMPLETED}`
    );
  });

  it('should throw when report is already FAILED', () => {
    // Arrange
    const failed = makeReport({ id: 5, status: ReportStatus.FAILED });
    const { service } = makeService([failed]);

    // Act & Assert
    expect(() => service.failAnalysis(5, 'err')).toThrow(
      `Cannot fail report 5 in terminal status: ${ReportStatus.FAILED}`
    );
  });

  it('should throw when report does not exist', () => {
    const { service } = makeService([]);
    expect(() => service.failAnalysis(999, 'err')).toThrow(
      'Analysis report not found with id: 999'
    );
  });
});
